dist/pythonlibs/riotctrl_shell/gnrc.py

```diff
@@ -0,0 +1,170 @@
+# Copyright (C) 2019-20 Freie Universität Berlin
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+"""
+GNRC-related shell interactions
+
+Defines GNRC-related shell command interactions
+"""
+
+import re
+
+from .base import ShellInteraction, ShellInteractionParser
+
+
+# ==== Parsers ====
+
+class GNRCICMPv6EchoParser(ShellInteractionParser):
+    @staticmethod
+    def _add_reply(res, reply):
+        reply["seq"] = int(reply["seq"])
+        reply["ttl"] = int(reply["ttl"])
+        reply["rtt"] = float(reply["rtt"])
+        if reply.get("rssi") is not None:
+            reply["rssi"] = int(reply["rssi"])
+        if "replies" in res:
+            res["replies"].append(reply)
+        else:
+            res["replies"] = [reply]
+
+    @staticmethod
+    def _set_stats(res, stats):
+        stats["packet_loss"] = int(stats["packet_loss"])
+        stats["rx"] = int(stats["rx"])
+        stats["tx"] = int(stats["tx"])
+        res["stats"] = stats
+
+    @staticmethod
+    def _set_rtts(res, rtts):
+        rtts["min"] = float(rtts["min"])
+        rtts["avg"] = float(rtts["avg"])
+        rtts["max"] = float(rtts["max"])
+        res["rtts"] = rtts
+
+    def parse(self, cmd_output):
+        res = {}
+        c_reply = re.compile(r"\d+ bytes from (?P<source>[0-9a-f:]+): "
+                             r"icmp_seq=(?P<seq>\d+) ttl=(?P<ttl>\d+) "
+                             r"(rssi=(?P<rssi>-?\d+) dBm )?"
+                             r"time=(?P<rtt>\d+.\d+) ms")
+        c_stats = re.compile(r"(?P<tx>\d+) packets transmitted, "
+                             r"(?P<rx>\d+) packets received, "
+                             r"(?P<packet_loss>\d+)% packet loss")
+        c_rtts = re.compile(r"round-trip min/avg/max = (?P<min>\d+.\d+)/"
+                            r"(?P<avg>\d+.\d+)/(?P<max>\d+.\d+) ms")
+        for line in cmd_output.splitlines():
+            if "stats" not in res:  # If final stats were not found yet
+                m = c_reply.match(line)
+                if m is not None:
+                    self._add_reply(res, m.groupdict())
+                    continue
+                m = c_stats.match(line)
+                if m is not None:
+                    self._set_stats(res, m.groupdict())
+                    continue
+            else:
+                m = c_rtts.match(line)
+                if m is not None:
+                    self._set_rtts(res, m.groupdict())
+                    return res
+        # Unable to parse RTTs, so something went wrong.
+        return None
+
+
+class GNRCPktbufStatsParser(ShellInteractionParser):
+    class Results(dict):
```

Comment:
Question: what is the benefit of doing this (nested class which inherits a dict) other than looking overly complex ?

Reply:
The main motivation is having useful methods provided for operations in the shell command people usually don't understand immediately (yes! I _always_ get questions about this when we doing testing):

- determine if the packet buffer is empty
- determine how much of the packet buffer was used in total

As to why it is nested: Except for read operations, this class is not useful outside of the parser class, and the setters should be kept private to its outer class (which arguably, I could move there, as with the ping parser, but they are more related to the nested class, so I put them there).

As to why inheriting dict, see https://github.com/RIOT-OS/RIOT/pull/11406#discussion_r449481667: You asked for having it simple in Ping so when we want methods here _and_ consistency it should also be a dict-like.

Reply:
Alternatively, we could have items `'is_empty'` and `'fullest_capacity'` in the dictionary, but since they are dependent on the values of other items, this would become even more complicated when those are changed and tbh weirder than a nested class inheriting from dict to provide item-dependent methods.

Reply:
But all this could be done without using a nested class anyway. The private methods of Results (used directly from the parser btw), could be private method of the parser.
I mean when things can be done a simpler way, there's no reason to overdesign them.

Reply:
Fine... as I can't convince you, I bow to your will :-P